File: 06-测试/10-测试-赋予武器-全模式/weapon_id_finder.py
import frida
import sys
import json
import time

class WeaponIDFinder:
    def __init__(self):
        self.session = None
        self.script = None
        self.weapon_log = []
        
    def connect(self, process_name="UnityCrossFire.exe"):
        try:
            self.session = frida.attach(process_name)
            print(f"[+] 已连接到进程: {process_name}")
            return True
        except Exception as e:
            print(f"[-] 连接失败: {e}")
            return False
    
    def load_script(self, script_path="weapon-id-finder.js"):
        try:
            with open(script_path, 'r', encoding='utf-8') as f:
                script_code = f.read()
            
            self.script = self.session.create_script(script_code)
            self.script.on('message', self.on_message)
            self.script.load()
            print(f"[+] 脚本已加载: {script_path}")
            return True
        except Exception as e:
            print(f"[-] 加载脚本失败: {e}")
            return False
    
    def on_message(self, message, data):
        if message['type'] == 'send':
            payload = message['payload']
            if payload['type'] == 'log':
                level = payload.get('level', 'info')
                module = payload.get('module', 'Unknown')
                msg = payload.get('message', '')
                print(f"[{level.upper()}][{module}] {msg}")
        elif message['type'] == 'error':
            print(f"[ERROR] {message['description']}")
    
    def hook_give_weapon(self):
        try:
            result = self.script.exports.hook_give_weapon()
            return result
        except Exception as e:
            print(f"[-] Hook失败: {e}")
            return False
    
    def get_weapon_name(self, weapon_id):
        try:
            name = self.script.exports.get_weapon_name(weapon_id)
            return name
        except Exception as e:
            return None
    
    def test_weapon_range(self, start_id, end_id):
        try:
            results = self.script.exports.test_weapon_range(start_id, end_id)
            return results
        except Exception as e:
            print(f"[-] 测试失败: {e}")
            return []
    
    def get_weapon_log(self):
        try:
            log = self.script.exports.get_weapon_log()
            return log
        except Exception as e:
            print(f"[-] 获取日志失败: {e}")
            return []
    
    def clear_weapon_log(self):
        try:
            self.script.exports.clear_weapon_log()
            print("[+] 武器日志已清空")
        except Exception as e:
            print(f"[-] 清空日志失败: {e}")

def main():
    print("=" * 60)
    print("武器ID查找工具")
    print("=" * 60)
    
    finder = WeaponIDFinder()
    
    if not finder.connect():
        return
    
    if not finder.load_script():
        return
    
    print("\n使用说明:")
    print("1. 选择 'Hook GiveWeapon' 开始监听")
    print("2. 在游戏中正常选择武器")
    print("3. 返回查看记录的武器ID")
    print("4. 或使用批量测试功能")
    
    while True:
        print("\n" + "=" * 60)
        print("选项:")
        print("  1. Hook GiveWeapon 方法")
        print("  2. 查看武器日志")
        print("  3. 清空武器日志")
        print("  4. 批量测试武器ID (1-100)")
        print("  5. 自定义范围测试")
        print("  6. 查询单个武器名称")
        print("  7. 导出武器ID列表")
        print("  0. 退出")
        print("=" * 60)
        
        choice = input("请选择: ").strip()
        
        if choice == '1':
            result = finder.hook_give_weapon()
            if result:
                print("[+] Hook成功，现在可以在游戏中选择武器")
                print("[+] 返回此界面查看记录的武器ID")
            else:
                print("[-] Hook失败")
        
        elif choice == '2':
            log = finder.get_weapon_log()
            if log:
                print("\n武器日志:")
                for entry in log:
                    weapon_id = entry.get('weaponIndex', 'Unknown')
                    time_str = entry.get('time', 'Unknown')
                    print(f"  [{time_str}] 武器ID: {weapon_id}")
            else:
                print("[-] 武器日志为空")
        
        elif choice == '3':
            finder.clear_weapon_log()
        
        elif choice == '4':
            print("\n开始批量测试武器ID (1-100)...")
            results = finder.test_weapon_range(1, 100)
            if results:
                print(f"\n找到 {len(results)} 个有效武器:")
                for weapon in results:
                    print(f"  ID {weapon['id']:3d}: {weapon['name']}")
            else:
                print("[-] 未找到有效武器")
        
        elif choice == '5':
            try:
                start_id = int(input("输入起始ID: "))
                end_id = int(input("输入结束ID: "))
                print(f"\n开始测试武器ID ({start_id}-{end_id})...")
                results = finder.test_weapon_range(start_id, end_id)
                if results:
                    print(f"\n找到 {len(results)} 个有效武器:")
                    for weapon in results:
                        print(f"  ID {weapon['id']:3d}: {weapon['name']}")
                else:
                    print("[-] 未找到有效武器")
            except ValueError:
                print("[-] 请输入有效的数字")
        
        elif choice == '6':
            try:
                weapon_id = int(input("输入武器ID: "))
                name = finder.get_weapon_name(weapon_id)
                if name:
                    print(f"[+] 武器名称: {name}")
                else:
                    print(f"[-] 未找到武器 ID={weapon_id}")
            except ValueError:
                print("[-] 请输入有效的数字")
        
        elif choice == '7':
            log = finder.get_weapon_log()
            if log:
                filename = f"weapon_id_list_{int(time.time())}.json"
                with open(filename, 'w', encoding='utf-8') as f:
                    json.dump(log, f, indent=2, ensure_ascii=False)
                print(f"[+] 武器ID列表已导出: {filename}")
            else:
                print("[-] 武器日志为空，无法导出")
        
        elif choice == '0':
            print("退出程序")
            break
        
        else:
            print("[-] 无效选项")

if __name__ == "__main__":
    main()

File: 06-测试/10-测试-赋予武器-全模式/weapon-id-finder.js
// ============================================================
// weapon-id-finder.js - 武器ID查找工具
//
// 功能：
//   1. Hook GiveWeapon 方法，记录武器ID
//   2. 遍历武器字典，列出所有武器
//   3. 测试武器ID范围，获取有效武器
//
// 使用方法：
//   1. 在游戏中正常选择武器
//   2. 脚本会自动记录武器ID
//   3. 使用 RPC 接口查询武器信息
// ============================================================

(function() {
    'use strict';

    // ==================== 1. 日志系统 ====================
    var MAX_LOGS_PER_MODULE = 100;
    var moduleLogCounts = {};
    var weaponLog = [];

    function log(level, module, message) {
        if (!moduleLogCounts[module]) moduleLogCounts[module] = 0;
        if (moduleLogCounts[module] >= MAX_LOGS_PER_MODULE) return;
        moduleLogCounts[module]++;
        var fullMsg = '[' + module + '] ' + message;
        console.log('[' + level + '] ' + fullMsg);
        try {
            send({type: 'log', level: level, module: module, message: message});
        } catch(e) {}
    }

    // ==================== 2. 模块查找 ====================
    var _gameAssembly = null;

    function getGameAssembly() {
        if (_gameAssembly) return _gameAssembly;
        try {
            var mod = Process.findModuleByName('GameAssembly.dll');
            if (!mod) {
                log('error', '系统', '未找到 GameAssembly.dll');
                return null;
            }
            _gameAssembly = mod;
            log('success', '系统', 'GameAssembly.dll 基址: ' + mod.base);
            return mod;
        } catch(e) {
            log('error', '系统', '获取模块失败: ' + e.message);
            return null;
        }
    }

    // ==================== 3. RVA地址常量 ====================
    var RVA = {
        GameManager_TypeInfo: 0x0E2933C,
        GiveWeapon: 0xAFB390,
        GetWpnName: 0xAFB1F0,
        GetWpnData: 0xAFB130,
    };

    // ==================== 4. 字段偏移常量 ====================
    var OFF = {
        Klass_staticFields: 0x5C,
        GM_WpnDictionary: 0x18,
        GM_WpnDataDictionary: 0x1C,
        WD_wpnIndex: 0xC,
        WD_weaponName: 0x14,
    };

    // ==================== 5. 安全读取函数 ====================
    function readPtr(addr) {
        try {
            if (!addr || addr.isNull()) return null;
            var v = addr.readPointer();
            return (v && !v.isNull()) ? v : null;
        } catch(e) {
            return null;
        }
    }

    function readS32(addr) {
        try {
            if (!addr || addr.isNull()) return null;
            return addr.readS32();
        } catch(e) {
            return null;
        }
    }

    function readUtf8String(addr, length) {
        try {
            if (!addr || addr.isNull()) return null;
            return addr.readUtf8String(length);
        } catch(e) {
            return null;
        }
    }

    // ==================== 6. Hook GiveWeapon 方法 ====================
    function hookGiveWeapon() {
        try {
            var mod = getGameAssembly();
            if (!mod) return false;
            var base = mod.base;

            var giveWeaponAddr = base.add(RVA.GiveWeapon);
            
            Interceptor.attach(giveWeaponAddr, {
                onEnter: function(args) {
                    var player = args[0];
                    var weaponIndex = args[1].toInt32();
                    var autoGiveUp = args[2].toInt32();
                    var autoSelect = args[3].toInt32();
                    
                    var timestamp = new Date().toISOString();
                    var logEntry = {
                        time: timestamp,
                        weaponIndex: weaponIndex,
                        autoGiveUp: autoGiveUp,
                        autoSelect: autoSelect
                    };
                    
                    weaponLog.push(logEntry);
                    
                    log('info', 'Hook', 'GiveWeapon 被调用');
                    log('info', 'Hook', '  武器ID: ' + weaponIndex);
                    log('info', 'Hook', '  自动放弃: ' + (autoGiveUp ? '是' : '否'));
                    log('info', 'Hook', '  自动选择: ' + (autoSelect ? '是' : '否'));
                    
                    var weaponName = getWeaponName(weaponIndex);
                    if (weaponName) {
                        log('success', 'Hook', '  武器名称: ' + weaponName);
                    }
                }
            });
            
            log('success', 'Hook', '已 Hook GiveWeapon 方法');
            return true;
            
        } catch(e) {
            log('error', 'Hook', 'Hook GiveWeapon 失败: ' + e.message);
            return false;
        }
    }

    // ==================== 7. 获取武器名称 ====================
    function getWeaponName(weaponIndex) {
        try {
            var mod = getGameAssembly();
            if (!mod) return null;
            var base = mod.base;

            var getWpnNameFunc = new NativeFunction(
                base.add(RVA.GetWpnName),
                'pointer',
                ['int']
            );

            var namePtr = getWpnNameFunc(weaponIndex);
            if (!namePtr || namePtr.isNull()) {
                return null;
            }

            return readUtf8String(namePtr);

        } catch(e) {
            return null;
        }
    }

    // ==================== 8. 批量测试武器ID ====================
    function testWeaponRange(startId, endId) {
        try {
            var results = [];
            
            log('info', 'Test', '开始测试武器ID范围: ' + startId + ' - ' + endId);
            
            for (var i = startId; i <= endId; i++) {
                var name = getWeaponName(i);
                if (name) {
                    results.push({
                        id: i,
                        name: name
                    });
                    log('success', 'Test', 'ID ' + i + ': ' + name);
                }
            }
            
            log('info', 'Test', '测试完成，找到 ' + results.length + ' 个有效武器');
            return results;
            
        } catch(e) {
            log('error', 'Test', '测试异常: ' + e.message);
            return [];
        }
    }

    // ==================== 9. 获取武器日志 ====================
    function getWeaponLog() {
        return weaponLog;
    }

    // ==================== 10. 清空武器日志 ====================
    function clearWeaponLog() {
        weaponLog = [];
        log('info', 'Log', '武器日志已清空');
    }

    // ==================== 11. 遍历武器字典 ====================
    function listWeaponsFromDict() {
        try {
            var mod = getGameAssembly();
            if (!mod) return null;
            var base = mod.base;

            var typeInfoSlot = base.add(RVA.GameManager_TypeInfo);
            var klass = readPtr(typeInfoSlot);
            if (!klass) {
                log('error', 'Dict', '无法获取 GameManager klass');
                return null;
            }

            var staticFields = readPtr(klass.add(OFF.Klass_staticFields));
            if (!staticFields) {
                log('error', 'Dict', '无法获取 GameManager 静态字段');
                return null;
            }

            var wpnDict = readPtr(staticFields.add(OFF.GM_WpnDictionary));
            if (!wpnDict) {
                log('error', 'Dict', 'WpnDictionary 为空');
                return null;
            }

            log('info', 'Dict', '武器字典地址: ' + wpnDict);
            log('info', 'Dict', '提示: 需要进一步分析字典结构');
            
            return wpnDict.toString();

        } catch(e) {
            log('error', 'Dict', '遍历武器字典异常: ' + e.message);
            return null;
        }
    }

    // ==================== 12. RPC接口导出 ====================
    rpc.exports = {
        hook_give_weapon: hookGiveWeapon,
        get_weapon_name: getWeaponName,
        test_weapon_range: testWeaponRange,
        get_weapon_log: getWeaponLog,
        clear_weapon_log: clearWeaponLog,
        list_weapons_from_dict: listWeaponsFromDict,
    };

    // ==================== 13. 初始化完成 ====================
    log('success', '系统', '武器ID查找工具已加载');
    log('info', '系统', '可用命令:');
    log('info', '系统', '  - hook_give_weapon()         Hook GiveWeapon 方法');
    log('info', '系统', '  - get_weapon_name(id)        获取武器名称');
    log('info', '系统', '  - test_weapon_range(start, end)  批量测试武器ID');
    log('info', '系统', '  - get_weapon_log()           获取武器日志');
    log('info', '系统', '  - clear_weapon_log()         清空武器日志');
    log('info', '系统', '  - list_weapons_from_dict()   遍历武器字典');
    log('info', '系统', '');
    log('info', '系统', '使用方法:');
    log('info', '系统', '  1. 调用 hook_give_weapon() 开始监听');
    log('info', '系统', '  2. 在游戏中正常选择武器');
    log('info', '系统', '  3. 调用 get_weapon_log() 查看记录的武器ID');
    log('info', '系统', '  4. 或使用 test_weapon_range(1, 100) 批量测试');

})();
